$MirOS: ports/net/djbdns/patches/patch-src_axfrdns_c,v 1.1.201.1 2005/05/14 23:21:02 tg Exp $
--- src/axfrdns.c.orig	Sun Feb 11 21:11:23 2001
+++ src/axfrdns.c	Wed Jul 26 23:30:05 2006
@@ -5,6 +5,9 @@
 #include "uint32.h"
 #include "uint16.h"
 #include "ip4.h"
+#ifdef DJB_V6ONLY
+#include "ip6.h"
+#endif /* DJB_V6ONLY */
 #include "tai.h"
 #include "buffer.h"
 #include "timeoutread.h"
@@ -123,7 +126,11 @@ void get(char *buf,unsigned int len)
   }
 }
 
+#ifndef DJB_V6ONLY
 char ip[4];
+#else
+char ip[16];
+#endif /* DJB_V6ONLY */
 unsigned long port;
 char clientloc[2];
 
@@ -237,6 +244,7 @@ void doaxfr(char id[2])
   byte_zero(clientloc,2);
   key[0] = 0;
   key[1] = '%';
+#ifndef DJB_V6ONLY
   byte_copy(key + 2,4,ip);
   r = cdb_find(&c,key,6);
   if (!r) r = cdb_find(&c,key,5);
@@ -246,6 +254,19 @@ void doaxfr(char id[2])
   if (r == -1) die_cdbread();
   if (r && (cdb_datalen(&c) == 2))
     if (cdb_read(&c,clientloc,2,cdb_datapos(&c)) == -1) die_cdbread();
+#else
+  if (byte_equal(ip,12,V4mappedprefix)) {
+    byte_copy(key + 2,4,ip+12);
+    r = cdb_find(&c,key,6);
+    if (!r) r = cdb_find(&c,key,5);
+    if (!r) r = cdb_find(&c,key,4);
+    if (!r) r = cdb_find(&c,key,3);
+    if (!r) r = cdb_find(&c,key,2);
+    if (r == -1) die_cdbread();
+    if (r && (cdb_datalen(&c) == 2))
+      if (cdb_read(&c,clientloc,2,cdb_datapos(&c)) == -1) die_cdbread();
+  }
+#endif /* DJB_V6ONLY */
 
   cdb_findstart(&c);
   for (;;) {
@@ -328,10 +349,17 @@ int main()
   axfr = env_get("AXFR");
   
   x = env_get("TCPREMOTEIP");
+#ifdef DJB_V6ONLY
+  if (x && ip6_scan(x,ip))
+    ;
+  else
+    byte_zero(ip,16);
+#else
   if (x && ip4_scan(x,ip))
     ;
   else
     byte_zero(ip,4);
+#endif /* DJB_V6ONLY */
 
   x = env_get("TCPREMOTEPORT");
   if (!x) x = "0";
@@ -357,7 +385,7 @@ int main()
 
     qlog(ip,port,header,zone,qtype," ");
 
-    if (byte_equal(qtype,2,DNS_T_AXFR)) {
+    if (byte_equal(qtype,2,DNS_T_AXFR) || byte_equal(qtype,2,DNS_T_IXFR)) {
       case_lowerb(zone,zonelen);
       fdcdb = open_read("data.cdb");
       if (fdcdb == -1) die_cdbread();
